File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/page/storage/Page.kt
package com.crescentflare.piratesgame.page.storage

import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.piratesgame.infrastructure.coreextensions.md5
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.lang.Exception

/**
 * Page storage: a single page item
 */
class Page {

    // --
    // Members
    // --

    var loadedData = emptyMap<String, Any>()
    val hash: String
    private val mapUtil = InflatorMapUtil()


    // --
    // Initialization
    // --

    constructor(jsonString: String) {
        var resultHash = "unknown"
        val type = object : TypeToken<Map<String, Any>>() {
        }.type
        try {
            val result = Gson().fromJson<Map<String, Any>>(jsonString, type)
            if (result != null) {
                loadedData = result
                resultHash = jsonString.md5()
            }
        } catch (ignored: Exception) {
            // No implementation
        }
        hash = resultHash
    }

    constructor(map: Map<String, Any>, hash: String = "unknown") {
        loadedData = map
        this.hash = hash
    }


    // --
    // Extract data
    // --

    val modules: List<Map<String, Any>>?
        get() {
            return optionalObjectMapList(loadedData, "modules")
        }

    val layout: Map<String, Any>?
        get() {
            val dataSetMap = mapUtil.asStringObjectMap(loadedData["dataSets"])
            if (dataSetMap != null) {
                val layout = mapUtil.asStringObjectMap(dataSetMap["layout"])
                if (layout != null) {
                    return layout
                }
            }
            return null
        }


    // --
    // Helper
    // --

    @Suppress("unchecked_cast")
    private fun optionalObjectMapList(map: Map<String, Any>, key: String): List<Map<String, Any>>? {
        val objectList = mapUtil.optionalObjectList(map, key)
        return if (objectList.size > 0 && mapUtil.asStringObjectMap(objectList[0]) == null) null else objectList as List<Map<String, Any>>
    }

}

File: pirates-ios/PiratesGame/Page/Storage/Page.swift
//
//  Page.swift
//  Page storage: a single page item
//

import Foundation

class Page {
    
    // --
    // MARK: Members
    // --

    var loadedData = [String: Any]()
    let hash: String
    

    // --
    // MARK: Initialization
    // --
    
    init(jsonString: String) {
        var resultHash = "unknown"
        loadedData = [:]
        if let jsonData = jsonString.data(using: .utf8) {
            if let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments) {
                if let parsedData = json as? [String: Any] {
                    loadedData = parsedData
                    resultHash = jsonString.md5()
                }
            }
        }
        hash = resultHash
    }
    
    init(jsonData: Data) {
        var resultHash = "unknown"
        loadedData = [:]
        if let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments) {
            if let parsedData = json as? [String: Any] {
                loadedData = parsedData
                resultHash = jsonData.md5()
            }
        }
        hash = resultHash
    }

    init(dictionary: [String: Any], hash: String = "unknown") {
        loadedData = dictionary
        self.hash = hash
    }


    // --
    // MARK: Extract data
    // --
    
    var modules: [[String: Any]]? {
        get {
            return loadedData["modules"] as? [[String: Any]]
        }
    }

    var layout: [String: Any]? {
        get {
            if let dataSets = loadedData["dataSets"] as? [String: Any] {
                return dataSets["layout"] as? [String: Any]
            }
            return nil
        }
    }

}
